### ISSUE TITLE
satpiler::init can be accidentally quadratic

### ISSUE THREAD

**Issue by horenmar** [master-keying/mks#23]
The behaviour of data resizing in `satpiler::init`, shown in the code below, can be easily made accidentally quadratic. 
```cpp
            if (idx >= data.size())
                data.resize(1+idx);
            data[idx] = lit;
```

This is similar to what `suggest::key` did in issue #8

**Comment by cernoch** [comment]
How should this piece of code look like? [A discussion on StackOverflow](https://stackoverflow.com/questions/37196329/c-standard-vector-resize-function#37197023)  suggests:
```cpp
            if (idx >= data.size()) {
                data.reserve(1+idx);
                data.resize(1+idx);
            }
            data[idx] = lit;
```

It's hard to believe this is the “proper” way...

**Comment by horenmar** [comment]
Well, [this](https://github.com/master-keying/mks/commit/b8c06f4d26f989789886cd94c012b3db1d4000ad#diff-8de5bd28f8320fc7b0f0118e1ab88ff4L51) is how it was solved in #8.

The trick is that as an inserting function, `data::insert` has to obey the growth curve.

**Comment by horenmar** [comment]
e4ed512bc removed the offending code from `satpiler` and moved it into `pd_matrix`. The code in `pd_matrix` does the resize properly via `vector::insert(Iter, Size, T)`.
